### PR TITLE
feat(participants-pane) - ParticipantsPane native component

### PR DESCRIPTION
**Pull request by Calinteodor**

### CODE REVIEW

File: react/features/video-menu/actions.any.js [hunk 1]
```diff
@@ -98,16 +98,33 @@ export function muteAllParticipants(exclude: Array<string>, mediaType: MEDIA_TYP
...
+    return (dispatch: Dispatch<any>) => {
         const knockingParticipantsIds = knockingParticipants.map(participant => participant.id);
 
         knockingParticipantsIds
             .map(id => lobbyEnabled && setKnockingParticipantApproval(id, true))
             .map(dispatch);
     };
 }
+
+
```
Comment (Calinteodor) [inline react/features/video-menu/actions.any.js:116]: This is something that I still got to work on.

File: react/features/video-menu/actions.any.js [hunk 2]
```diff
@@ -98,16 +98,33 @@ export function muteAllParticipants(exclude: Array<string>, mediaType: MEDIA_TYP
...
+/**
+ * Don't allow participants to unmute.
+ *
+ * @returns {Function}
+ */
+export function dontAllowUnmute() {
```
Comment (Calinteodor) [inline react/features/video-menu/actions.any.js:122]: This is something in progress.

File: react/features/participants-pane/components/native/MeetingParticipantItem.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    const isAudioMuted = useSelector(getIsParticipantAudioMuted(p));
+    const isVideoMuted = useSelector(getIsParticipantVideoMuted(p));
+    const openContextMenuDetails = useCallback(() => dispatch(showContextMenuDetails(p), [ dispatch ]));
+
+    return (
+        <ParticipantItem
+            audioMuteState = { isAudioMuted ? MediaState.Muted : MediaState.Unmuted }
+            isKnockingParticipant = { false }
+            name = { p.name }
+            onPress = { !p.local && openContextMenuDetails }
```
Comment (tmoldovan8x8) [inline react/features/participants-pane/components/native/MeetingParticipantItem.js:35]: There is a warning thrown I think on this onPress

File: react/features/participants-pane/components/native/ContextMenuMore.js [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    const cancel = useCallback(() => dispatch(hideDialog()), [ dispatch ]);
+    const muteEveryoneVideo = useCallback(() => dispatch(muteAllParticipants(exclude, MEDIA_TYPE.VIDEO)), [ dispatch ]);
+    const { t } = useTranslation();
+
+    return (
+        <BottomSheet
+            onCancel = { cancel }
+            style = { styles.contextMenuMore }>
+            <TouchableOpacity
+                onPress = { muteEveryoneVideo }
```
Comment (tmoldovan8x8) [inline react/features/participants-pane/components/native/ContextMenuMore.js:42]: This throws an error, does not seem to work

File: react/features/participants-pane/components/native/ContextMenuMore.js [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+                style = { styles.contextMenuItem }>
+                <Icon
+                    size = { 20 }
+                    src = { IconVideoOff } />
+                <Text style = { styles.contextMenuItemText }>{t('participantsPane.actions.stopEveryonesVideo')}</Text>
+            </TouchableOpacity>
+            <TouchableOpacity
+                onPress = { blockAudioVideo }
```
Comment (tmoldovan8x8) [inline react/features/participants-pane/components/native/ContextMenuMore.js:50]: Also this

File: react/features/participants-pane/components/native/MeetingParticipantList.js [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+            {
+                participants.map(p => (
+                    <MeetingParticipantItem
+                        key = { p.id }
+                        participant = { p } />)
+                )
+            }
+        </View>
+    );
+};
```
Comment (tmoldovan8x8) [inline react/features/participants-pane/components/native/MeetingParticipantList.js:56]: Could you please check the behaviour while screen rotating on Android. I've seen some strange UI on this component, but might be on others, as well

**Review by tmoldovan8x8 — CHANGES_REQUESTED** [review]

File: react/features/participants-pane/components/native/ParticipantItem.js [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+    /**
+     * The name of the participant. Used for showing lobby names.
+     */
+    name?: string,
+
+    /**
+     * Callback for when the mouse leaves this component
+     */
+    onLeave?: Function,
```
Comment (vp8x8) [inline react/features/participants-pane/components/native/ParticipantItem.js:46]: I don't think we need this on mobile.

File: react/features/participants-pane/components/native/ParticipantItem.js [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+    videoMuteState = MediaState.None
+}: Props) {
+
+    const displayName = name || useSelector(getParticipantDisplayNameWithId(p.id));
+    const { t } = useTranslation();
+
+    return (
+        <View style = { styles.participantContainer } >
+            <TouchableOpacity
+                /* eslint-disable-next-line react/jsx-no-bind */
```
Comment (vp8x8) [inline react/features/participants-pane/components/native/ParticipantItem.js:85]: I see no bind here. Is this comment necessary?

File: react/features/video-menu/actions.any.js [hunk 9]
```diff
@@ -92,3 +94,40 @@ export function muteAllParticipants(exclude: Array<string>, mediaType: MEDIA_TYP
...
+ * Don't allow participants to unmute video/audio.
+ *
+ * @returns {Function}
+ */
+export function blockParticipantsAudioVideo() {
+    return (dispatch: Dispatch<any>, getState: Function) => {
+        const state = getState();
+        const participants = state['features/base/participants'];
+
+        console.log(participants);
```
Comment (vp8x8) [inline react/features/video-menu/actions.any.js:127]: I don't think we need this log.

File: react/features/video-menu/components/native/VolumeSlider.js [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+     */
+    volumeLevel: number
+};
+
+/**
+ * Component that renders the volume slider.
+ *
+ * @returns {React$Element<any>}
+ */
+class VolumeSlider extends Component<Props, State> {
```
Comment (vp8x8) [inline react/features/video-menu/components/native/VolumeSlider.js:53]: Any reason for creating a class component instead of a functional one?

**Comment by vp8x8** [comment]
This needs a rebase.

File: react/features/participants-pane/components/native/MeetingParticipantItem.js (hunk 3 above)
Comment (Calinteodor) [reply react/features/participants-pane/components/native/MeetingParticipantItem.js]: Fixed.

File: react/features/participants-pane/components/native/ContextMenuMore.js (hunk 4 above)
Comment (Calinteodor) [reply react/features/participants-pane/components/native/ContextMenuMore.js]: Fixed.

File: react/features/base/react/components/native/HeaderWithNavigation.js [hunk 11]
```diff
@@ -48,17 +53,22 @@ class HeaderWithNavigation extends Component<Props> {
...
         return (
-            <Header>
-                { onPressBack && <BackButton onPress = { onPressBack } /> }
-                <HeaderLabel labelKey = { this.props.headerLabelKey } />
-                { onPressForward && <ForwardButton
-                    disabled = { this.props.forwardDisabled }
-                    labelKey = { this.props.forwardLabelKey }
-                    onPress = { onPressForward } /> }
-            </Header>
+            <>
```
[the remark this answers is not in the document]
Comment (Calinteodor) [reply react/features/base/react/components/native/HeaderWithNavigation.js]: Fixed.

File: react/features/conference/components/native/Conference.js [hunk 12]
```diff
@@ -71,6 +72,11 @@ type Props = AbstractProps & {
      */
     _fullscreenEnabled: boolean,
 
+    /**
+     * The indicator which determines if the participants pane is open.
+     */
+    _isOpen: boolean,
```
[the remark this answers is not in the document]
Comment (Calinteodor) [reply react/features/conference/components/native/Conference.js]: Renamed.

File: react/features/participants-pane/components/native/ContextMenuMeetingParticipantDetails.js [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+    /**
+     * Participant reference
+     */
+    participant: Object
+};
+
+export const ContextMenuMeetingParticipantDetails = ({ participant: p }: Props) => {
+    const [ volume, setVolume ] = useState(undefined);
+    const store = useStore();
+    const startSilent = store.getState['features/base/config'];
```
[the remark this answers is not in the document]
Comment (Calinteodor) [reply react/features/participants-pane/components/native/ContextMenuMeetingParticipantDetails.js]: Removed useStore and replaced with useSelector.

File: react/features/participants-pane/components/native/LobbyParticipantList.js [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+export const LobbyParticipantList = () => {
+    const {
+        lobbyEnabled,
+        knockingParticipants: participants
+    } = useSelector(getLobbyState);
+
+    const dispatch = useDispatch();
+    const admitAll = useCallback(() =>
+        dispatch(admitAllKnockingParticipants(participants, lobbyEnabled)),
+        [ dispatch ]);
```
[the remark this answers is not in the document]
Comment (Calinteodor) [reply react/features/participants-pane/components/native/LobbyParticipantList.js:24]: Replaced with admitMultiple action.

File: react/features/participants-pane/components/native/MeetingParticipantList.js [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+            <Text style = { styles.meetingListDescription }>
+                {t('participantsPane.headings.participantsList',
+                    { count: participants.length })}
+            </Text>
+            {
+                showInviteButton
+                && <Button
+                    children = { t('participantsPane.actions.invite') }
+                    /* eslint-disable-next-line react/jsx-no-bind */
+                    icon = { () =>
```
[the remark this answers is not in the document]
Comment (Calinteodor) [reply react/features/participants-pane/components/native/MeetingParticipantList.js:35]: This is react-native-paper framework specific. 
For other than material icons, I can only pass a function which returns a react element to be used an icon.

File: react/features/participants-pane/components/native/ParticipantItem.js (hunk 7 above)
Comment (Calinteodor) [reply react/features/participants-pane/components/native/ParticipantItem.js]: Removed.

File: react/features/participants-pane/components/native/ParticipantItem.js (hunk 8 above)
Comment (Calinteodor) [reply react/features/participants-pane/components/native/ParticipantItem.js]: Removed.

File: react/features/video-menu/components/native/VolumeSlider.js (hunk 10 above)
Comment (Calinteodor) [reply react/features/video-menu/components/native/VolumeSlider.js:53]: The only reason was that the components created for this feature are class based.

File: react/features/participants-pane/components/native/ContextMenuMeetingParticipantDetails.js [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+    const dispatch = useDispatch();
+    const cancel = useCallback(() => dispatch(hideDialog()), [ dispatch ]);
+    const changeVolume = useCallback(() => setVolume(volume), [ volume ]);
```
[the remark this answers is not in the document]
Comment (Calinteodor) [reply react/features/participants-pane/components/native/ContextMenuMeetingParticipantDetails.js]: If I rely only on the setVolume, I enter in an infinite render loop. 
A "Too many re-renders" error is thrown.

File: react/features/video-menu/components/native/VolumeSlider.js (hunk 10 above)
Comment (vp8x8) [reply react/features/video-menu/components/native/VolumeSlider.js:53]: Could you please use a functional component? The ideea is to try to use functional components wherever possible and hooks. So instead of inheritance we could reuse code through hooks.

Comment (damencho) [reply react/features/video-menu/components/native/VolumeSlider.js:53]: We need to discuss this. I'm not 100% sure that this is a good idea. There can be performance issues and re-rendering for those components, we were debugging the participants' pane the last few days ...

Comment (Calinteodor) [reply react/features/video-menu/components/native/VolumeSlider.js:47]: Updated

Comment (vp8x8) [reply react/features/video-menu/components/native/VolumeSlider.js:47]: @damencho There should be no difference in performance between class components & functional components. If there are multiple re-renders this must be because of how the component was built. I guess we should use React memo & Pure component more often.

Comment (damencho) [reply react/features/video-menu/components/native/VolumeSlider.js:47]: Yeah, I guess it was a bad implementation of the selectors as there were some places returning new objects (new constructed arrays) which will trigger new re-renders ... I'm not even sure where that was at the moment, we will soon create the PR.  
But there is a difference in how the comparison is done by react in the selectors it is === while in mapstatet2props it is deep equal comparison, so there is a big difference in this which can lead to re-renders.

Comment (damencho) [reply react/features/video-menu/components/native/VolumeSlider.js:47]: An example I just saw https://github.com/jitsi/jitsi-meet/blob/2174368d23e115781118b23ed6f00ce95ab30beb/react/features/av-moderation/functions.js#L95
This is used in a selector, this will be rendered on every action

Comment (damencho) [reply react/features/video-menu/components/native/VolumeSlider.js:47]: And maybe I added that, by the way :) not sure

File: react/features/participants-pane/components/native/MeetingParticipantList.js (hunk 6 above)
Comment (Calinteodor) [reply react/features/participants-pane/components/native/MeetingParticipantList.js:54]: Can you please attach a screenshot with this because I cannot reproduce it. Thanks.

File: react/features/participants-pane/components/native/ContextMenuMore.js (hunk 5 above)
Comment (Calinteodor) [reply react/features/participants-pane/components/native/ContextMenuMore.js:46]: This will be covered with the a-v moderation feature.

File: react/features/video-menu/components/native/VolumeSlider.js [hunk 17]
```diff
@@ -0,0 +1,164 @@
...
+ * @param {Object} state - The Redux state.
+ * @param {Object} ownProps - The own props of the component.
+ * @returns {Props}
+ */
+function mapStateToProps(state, ownProps): Object {
+    const { participant } = ownProps;
+    const { startSilent } = state['features/base/config'];
+    const { participantsVolume } = state['features/participants-pane'];
+    const getParticipant = participant.id
+        ? getParticipantById(state, participant.id) : getLocalParticipant(state);
```
Comment (vp8x8) [inline react/features/video-menu/components/native/VolumeSlider.js:154]: Why is this necessary? Don't we already have the `id` of the participant?

File: react/features/video-menu/components/native/VolumeSlider.js [hunk 18]
```diff
@@ -0,0 +1,164 @@
...
+    const { id } = getParticipant;
+
+    return {
+        _startSilent: Boolean(startSilent),
+        _volume: participant.local ? undefined : participantsVolume[id]
```
Comment (vp8x8) [inline react/features/video-menu/components/native/VolumeSlider.js:159]: You could create an separate selector for this.

**Comment by vp8x8** [comment]
This looks pretty good (functionality wise). A few remarks:

1. If a participant quits while the the participants pane & drawer are open for it, there is an error:
Screenshot 2021-07-02 at 12 11 52

2. There are a few design inconsistency:
* The upper right close button looks way to big. Please check the dimensions from the design and use the tokens.
* The same applies for the bottom buttons (Mute and ... button)
* The avatar  from the participant list and the icons from  actions menu looks like are not middle aligned.
* There are a few keys missing: `participantsPane.actions.mute` & `participantsPane.actions.muteEveryoneElse`
* The knob for the slider is way too big. 

3. Can we get rid of the warning regarding the Slider? It looks like it's deprecated and we have to use the one from react-native-community instead of the core one.

**Review by vp8x8 — CHANGES_REQUESTED** [review]

File: react/features/base/react/components/native/HeaderWithNavigation.js [hunk 19]
```diff
@@ -48,17 +53,18 @@ class HeaderWithNavigation extends Component<Props> {
...
         return (
-            <Header>
-                { onPressBack && <BackButton onPress = { onPressBack } /> }
-                <HeaderLabel labelKey = { this.props.headerLabelKey } />
-                { onPressForward && <ForwardButton
-                    disabled = { this.props.forwardDisabled }
-                    labelKey = { this.props.forwardLabelKey }
-                    onPress = { onPressForward } /> }
-            </Header>
+            !{ hideHeaderWithNavigation }
```
[the remark this answers is not in the document]
Comment (Calinteodor) [reply react/features/base/react/components/native/HeaderWithNavigation.js:59]: > This looks pretty good (functionality wise). A few remarks:
> 
> 1. If a participant quits while the the participants pane & drawer are open for it, there is an error:
> 
> Screenshot 2021-07-02 at 12 11 52
> 
> 1. There are a few design inconsistency:
> 
> * The upper right close button looks way to big. Please check the dimensions from the design and use the tokens.
> * The same applies for the bottom buttons (Mute and ... button)
> * The avatar  from the participant list and the icons from  actions menu looks like are not middle aligned.
> * There are a few keys missing: `participantsPane.actions.mute` & `participantsPane.actions.muteEveryoneElse`
> * The knob for the slider is way too big.
> 
> 1. Can we get rid of the warning regarding the Slider? It looks like it's deprecated and we have to use the one from react-native-community instead of the core one.

Everything is fixed except the knob. For the Slider component we have no thumbSize prop. 
There is an issue on the community that is still open
https://github.com/callstack/react-native-slider/issues/97.

File: react/features/video-menu/components/native/VolumeSlider.js (hunk 17 above)
Comment (Calinteodor) [reply react/features/video-menu/components/native/VolumeSlider.js:154]: Fixed.

**Comment by vp8x8** [comment]
3 more things:

1. The icons from the BottomSheet look too big. They're definitely bigger than 16px.
2. If a participant quits while BottomSheet is opened it remains opened with the respective actions. It should automatically close when a participant leaves.
3. I think you could test the android build using the community slider if you modify 2 files: `/android/settings.graddle`, and `android/sdk/build.gradle`. Make the adjustments similar to `netinfo` package.

**Comment by Calinteodor** [comment]
> 3 more things:
> 
> 1. The icons from the BottomSheet look too big. They're definitely bigger than 20px.
> 2. If a participant quits while BottomSheet is opened it remains opened with the respective actions. It should automatically close when a participant leaves.
> 3. I think you could test the android build using the community slider if you modify 2 files: `/android/settings.graddle`, and `android/sdk/build.gradle`. Make the adjustments similar to `netinfo` package.

Fixed 1 and 2. Thanks for the hint regarding 3, but it doesn't work after I updated those two files. 
Also, please don't merge this because @damencho and @hristoterezov change how the participants are stored and we should wait for their PR to be merged first. Thanks!

**Review by hristoterezov — APPROVED** [review]

**Comment by hristoterezov** [comment]
@vp8x8 I believe all change requests were addressed. Please let me know if I missed something so that @Calinteodor can fix it.
